src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/X509Certificates/X509SubjectAlternativeNameExtension.cs

```diff
@@ -0,0 +1,187 @@
+                ReadOnlySpan<char> match = hostname;
+
+                if (hostname.EndsWith('.'))
+                {
+                    match = match.Slice(0, match.Length - 1);
+
+                    if (match.IsEmpty)
+                    {
+                        return false;
+                    }
+                }
+
+                ReadOnlySpan<char> afterFirstDot = default;
+                int firstDot = match.IndexOf('.');
+
+                if (firstDot == 0)
+                {
+                    return false;
+                }
+
+                if (firstDot > 0)
+                {
+                    afterFirstDot = match.Slice(firstDot + 1);
+                }
+
+                foreach (string embedded in EnumerateDnsNames())
+                {
+                    if (embedded.Length == 0)
+                    {
+                        continue;
+                    }
+
+                    ReadOnlySpan<char> embeddedSpan = embedded;
+
+                    if (embedded.EndsWith('.'))
+                    {
+                        embeddedSpan = embeddedSpan.Slice(0, embeddedSpan.Length - 1);
+                    }
+
+                    if (embeddedSpan.StartsWith("*.") && embeddedSpan.Length > 2)
+                    {
+                        if (embeddedSpan.Slice(2).Equals(afterFirstDot, StringComparison.OrdinalIgnoreCase))
+                        {
+                            return true;
+                        }
+                    }
+                    else if (embeddedSpan.Equals(match, StringComparison.OrdinalIgnoreCase))
+                    {
+                        return true;
+                    }
+                }
+            }
+
+            return false;
+        }
+
+        public IEnumerable<string> EnumerateDnsNames()
+        {
+            List<GeneralNameAsn> decoded = (_decoded ??= Decode(RawData));
+
+            return EnumerateDnsNames(decoded);
+        }
+
+        private static IEnumerable<string> EnumerateDnsNames(List<GeneralNameAsn> decoded)
+        {
+            foreach (GeneralNameAsn item in decoded)
+            {
+                if (item.DnsName is not null)
+                {
+                    yield return item.DnsName;
+                }
+            }
+        }
+
+        public IEnumerable<IPAddress> EnumerateIPAddresses()
+        {
+            List<GeneralNameAsn> decoded = (_decoded ??= Decode(RawData));
+
+            return EnumerateIPAddresses(decoded);
+        }
+
+        private static IEnumerable<IPAddress> EnumerateIPAddresses(List<GeneralNameAsn> decoded)
+        {
+            foreach (GeneralNameAsn item in decoded)
+            {
+                if (item.IPAddress.HasValue)
+                {
+                    ReadOnlySpan<byte> value = item.IPAddress.GetValueOrDefault().Span;
+
+                    if (value.Length is 4 or 16)
+                    {
+                        yield return new IPAddress(value);
+                    }
```

Comment:
Is skipping malformed IP addresses reasonable? Should there be an else?

Reply:
Yeah, SAN does talk only about IPv4 and IPv6 here, so complaining about other values as an encoding error might be fair.